### ISSUE TITLE
linspace behave differently than numpy linspace

### ISSUE THREAD

**Issue by zli96** [pytorch/pytorch#88652]
### 🐛 Describe the bug

```python
import torch
import numpy as np
torch.linspace(1,32,10,dtype=int)
np.linspace(1,32,10,dtype=int)
```
Pytorch gives you: tensor([ 1,  4,  7, 11, 14, 17, 20, 24, 28, 32])
Numpy gives you: array([ 1,  4,  7, 11, 14, 18, 21, 25, 28, 32])

Maybe I should post this issue on numpy repository as well...

### Versions

Collecting environment information...
PyTorch version: 1.11.0a0+git7f4db2c
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Rocky Linux release 8.6 (Green Obsidian) (x86_64)
GCC version: (GCC) 8.4.0
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.28

Python version: 3.10.2 (main, Jun  8 2022, 10:30:57) [GCC 8.5.0 20210514 (Red Hat 8.5.0-10)] (64-bit runtime)
Python platform: Linux-4.18.0-372.19.1.el8_6.x86_64-x86_64-with-glibc2.28
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.931
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.2
[pip3] numpy-groupies==0.9.17
[pip3] numpydoc==1.2.1
[pip3] pytorch-lightning==1.7.7
[pip3] torch==1.11.0a0+git7f4db2c
[pip3] torchmetrics==0.10.0
[pip3] torchvision==0.12.0a0+7f4db2c
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.21.5           py39he7a7128_1  
[conda] numpy-base                1.21.5           py39hf524024_1  
[conda] numpydoc                  1.2                pyhd3eb1b0_0

**Comment by albanD** [comment]
cc @mruberry is this a known behavior?

**Comment by mruberry** [comment]
No; that's very odd. The issue appears to be specific to the CPU and int64 dtype:

```
torch.linspace(1, 32, 10, dtype=torch.int32)
: tensor([ 1,  4,  7, 11, 14, 18, 21, 25, 28, 32], dtype=torch.int32)

torch.linspace(1, 32, 10, device='cuda', dtype=torch.int32)
: tensor([ 1,  4,  7, 11, 14, 18, 21, 25, 28, 32], device='cuda:0', dtype=torch.int32)

np.linspace(1, 32, 10, dtype=np.int32)
: array([ 1,  4,  7, 11, 14, 18, 21, 25, 28, 32], dtype=int32)

# This is the odd one out
torch.linspace(1, 32, 10, dtype=torch.int64)
: tensor([ 1,  4,  7, 11, 14, 17, 20, 24, 28, 32])

torch.linspace(1, 32, 10, device='cuda', dtype=torch.int64)
: tensor([ 1,  4,  7, 11, 14, 18, 21, 25, 28, 32], device='cuda:0')

np.linspace(1, 32, 10, dtype=np.int64)
: array([ 1,  4,  7, 11, 14, 18, 21, 25, 28, 32])

```

**Comment by yanbing-j** [comment]
The scalar version can get the right answer, while vec implementation cannot. WIP.

**Comment by jgong5** [comment]
> The scalar version can get the right answer, while vec implementation cannot. WIP.

The root cause is that the vectorized implementation tends to suffer from the catastrophic cancellation of floating point arithemtic since both the base (`start + step * idx`) and the `step` are not exact. The scalar implementation is fine since `start` is always an integer and the result would be truncated to integer as well. Skipping the vectorized implementation makes sense as @yanbing-j added since the vec doesn't contribute to performance anyway.
https://github.com/pytorch/pytorch/pull/89048/files#diff-52f1a03f9142df1299737546cdb0e6dddf7e3140ca8f254e96eac9954fbddd4dL69
